packages/buidler-etherscan/src/solc/bytecode.ts

```diff
@@ -1,59 +1,155 @@
+import { Artifacts } from "@nomiclabs/buidler/plugins";
 import { RunTaskFunction } from "@nomiclabs/buidler/types";
+import fs from "fs-extra";
+import path from "path";
 
 import { METADATA_LENGTH_SIZE, readSolcMetadataLength } from "./metadata";
 import { InferralType } from "./version";
 
+type BytecodeComparison =
+  | { match: false }
+  | { match: true; contractInformation: BytecodeExtractedData };
+
+interface BytecodeExtractedData {
+  immutableValues: ImmutableValues;
+  libraryLinks: ResolvedLinks;
+  normalizedBytecode: string;
+}
+
+interface ResolvedLinks {
+  [sourceName: string]: {
+    [libraryName: string]: string;
+  };
+}
+
+interface ImmutableValues {
+  [key: string]: string;
+}
+
+interface BytecodeSlice {
+  start: number;
+  length: number;
+}
+
+type LinkReferences = CompilerOutputBytecode["linkReferences"][string][string];
+type NestedSliceReferences = BytecodeSlice[][];
+
+/* Taken from stack trace buidlerevm internals
+ *  This is not an exhaustive interface for compiler input nor output.
+ */
+
+interface CompilerInput {
+  language: "Solidity";
+  sources: { [sourceName: string]: { content: string } };
+  settings: {
+    optimizer: { runs: number; enabled: boolean };
+    evmVersion?: string;
+    libraries?: ResolvedLinks;
+  };
+}
+
+interface CompilerOutput {
+  sources: CompilerOutputSources;
+  contracts: {
+    [sourceName: string]: {
+      [contractName: string]: {
+        abi: any;
+        evm: {
+          bytecode: CompilerOutputBytecode;
+          deployedBytecode: CompilerOutputBytecode;
+          methodIdentifiers: {
+            [methodSignature: string]: string;
+          };
+        };
+      };
+    };
+  };
+}
+
+interface CompilerOutputSource {
+  id: number;
+  ast: any;
+}
+
+interface CompilerOutputSources {
+  [sourceName: string]: CompilerOutputSource;
+}
+
+interface CompilerOutputBytecode {
+  object: string;
+  opcodes: string;
+  sourceMap: string;
+  linkReferences: {
+    [sourceName: string]: {
+      [libraryName: string]: Array<{ start: 0; length: 20 }>;
+    };
+  };
+  immutableReferences?: {
+    [key: string]: Array<{ start: number; length: number }>;
+  };
+}
+
+interface BuildInfo {
+  input: CompilerInput;
+  output: CompilerOutput;
+  solcVersion: string;
+}
+
+interface ContractBuildInfo extends BuildInfo {
```

Comment:
I feel like this shouldn't be an extension, but a composition. Something like:

```
sourceName: string;
contractName: string;
buildInfo: BuildInfo
```

Not sure about the name though. Maybe `ContractBuildInfo` is still fine, idk.